supernode/daemon/mgr/pieceerror/manager.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright The Dragonfly Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package pieceerror
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/dragonflyoss/Dragonfly/apis/types"
+	"github.com/dragonflyoss/Dragonfly/pkg/errortypes"
+	"github.com/dragonflyoss/Dragonfly/pkg/syncmap"
+	"github.com/dragonflyoss/Dragonfly/supernode/config"
+	"github.com/dragonflyoss/Dragonfly/supernode/daemon/mgr"
+
+	"github.com/sirupsen/logrus"
+)
+
+var _ mgr.PieceErrorMgr = &Manager{}
+
+const (
+	ErrHandlerChanSize = 100
+)
+
+// TODO: make it pluggable
+type Handler interface {
+	Handle(ctx context.Context, pieceErrorRequest *types.PieceErrorRequest) error
+}
+
+type Manager struct {
+	cfg *config.Config
+
+	gcManager  mgr.GCMgr
+	cdnManager mgr.CDNMgr
+
+	// error handler
+	pieceErrChan       chan *types.PieceErrorRequest
+	errorHandlingStore *syncmap.SyncMap
+}
+
+func NewManager(cfg *config.Config, gcManager mgr.GCMgr, cdnManager mgr.CDNMgr) (*Manager, error) {
+	return &Manager{
+		cfg:          cfg,
+		gcManager:    gcManager,
+		cdnManager:   cdnManager,
+		pieceErrChan: make(chan *types.PieceErrorRequest, ErrHandlerChanSize),
+	}, nil
+}
+
+// HandlePieceError the peer should report the error with related info when
+// it failed to download a piece from supernode.
+// And the supernode should handle the piece Error and do some repair operations.
+func (em *Manager) HandlePieceError(ctx context.Context, pieceErrorRequest *types.PieceErrorRequest) error {
+	// ignore the error that isn't caused by downloading from supernode
+	if em.cfg.IsSuperPID(pieceErrorRequest.DstPid) {
+		return nil
+	}
+
+	// if the error is handling, we should ignore it.
+	if _, err := em.errorHandlingStore.Get(pieceErrorRequest.TaskID); err == nil {
```

Comment:
I don't see the step to add taskId into `errorHandlingStore`. And there's no step to remove the taskId from `errorHandlingStore` after handling.